File: common/modules/debrid/route.js
// Pure debrid policy, free of UI imports so it can be tested under plain Node.
// Two decisions live here: how a play request is routed (debrid only mode must never
// reach the torrent client, whatever the input), and which search results are listed.
import { Availability, streamsInstantly } from './availability.js'

const magnetRx = /^magnet:.*urn:btih:[a-f\d]{40}/i
const hexRx = /^[a-f\d]{40}$/i

/** @param {any} torrentID */
function usable (torrentID) {
  return (typeof torrentID === 'string' && (magnetRx.test(torrentID) || hexRx.test(torrentID)) && torrentID) || null
}

/**
 * Decides how a play request should be handled.
 * @param {Object} options
 * @param {any} options.torrentID - Magnet URI, info hash, .torrent link, or torrent file bytes.
 * @param {any} [options.hash] - Info hash when known, used when the link itself is not resolvable.
 * @param {boolean} options.serviceSelected - A debrid service is selected in settings.
 * @param {boolean} options.serviceReady - The service has an API key configured.
 * @param {boolean} options.offline - The client has no network connection.
 * @param {string} options.mode - The debridMode setting, 'prefer' or 'only'.
 * @returns {{ action: 'torrent', only: boolean } | { action: 'block', reason: 'key' | 'offline' | 'source', only: boolean } | { action: 'resolve', id: string, only: boolean }}
 *   `only` reports whether debrid only mode governs this decision, so callers never
 *   have to re-derive the fallback rules that were already applied here.
 */
export function routeDebrid ({ torrentID, hash, serviceSelected, serviceReady, offline, mode }) {
  // with no service selected debrid is entirely out of the picture, only mode included
  if (!serviceSelected) return { action: 'torrent', only: false }
  const only = mode === 'only'
  if (!serviceReady) return only ? { action: 'block', reason: 'key', only } : { action: 'torrent', only }
  if (offline) return only ? { action: 'block', reason: 'offline', only } : { action: 'torrent', only }
  const id = usable(torrentID) || usable(hash)
  if (!id) return only ? { action: 'block', reason: 'source', only } : { action: 'torrent', only }
  return { action: 'resolve', id, only }
}

/**
 * The API key stored for a debrid service.
 *
 * Every service keeps its own, so switching between them in settings swaps the key rather than
 * losing it, and an account on one service can never be addressed with another's key. The
 * settings tab and the playback module both read through here so they cannot disagree about
 * which key is in play.
 * @param {{ debridApiKeys?: Record<string, string> }} settings
 * @param {string} [service] - Service id, defaulting to the selected one.
 * @returns {string} Empty when that service has no key yet.
 */
export function debridKey (settings, service = settings?.debridService) {
  return (service && settings?.debridApiKeys?.[service]) || ''
}

/**
 * Whether a search result belongs in the listed results rather than the hidden ones.
 *
 * With no filters on this is upstream's rule, widened only by the fact that a cached release
 * streams without seeders. The two debrid cases that narrow it:
 * - the cached filter shows nothing but confirmed cached releases;
 * - debrid only mode hides releases the service has said it cannot serve, since with no
 *   torrent client to fall back on those cannot play at all.
 *
 * An *available* release is deliberately not widened in: the service would still have to pull
 * it from the swarm, so a seederless one is no more playable than it was without debrid.
 * @param {{ seeders?: number, source?: { managed?: boolean } }} result
 * @param {string} [availability] - What the service said about this release.
 * @param {{ cachedOnly?: boolean, only?: boolean }} [options] - The debrid filters in force.
 * @returns {boolean}
 */
export function listResult (result, availability, { cachedOnly, only } = {}) {
  const cached = streamsInstantly(availability)
  if (cachedOnly) return cached
  if (only && availability === Availability.UNAVAILABLE) return false
  return result?.seeders > 0 || Boolean(result?.source?.managed) || cached
}

File: crates/core/src/route.rs
//! How a play request is routed between the local torrent engine and a debrid
//! service. Port of common/modules/debrid/route.js; test/unit/debrid/route.test.js
//! is the behavioural reference.

use serde::{Deserialize, Serialize};
use shiru_domain::parse_hash;

/// The debridMode setting.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DebridMode {
    /// Debrid First: cached releases stream from the service, anything uncached
    /// falls back to torrents.
    Prefer,
    /// Debrid Only: playback always uses the service, torrents never start.
    Only,
}

/// Why playback was blocked instead of routed.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum BlockReason {
    /// Debrid only mode is on but no API key is set.
    Key,
    /// The client is offline, so the service cannot be reached.
    Offline,
    /// The source only provides something debrid cannot resolve (no usable hash).
    Source,
}

#[derive(Debug, Clone, Default)]
pub struct RouteInput<'a> {
    /// Magnet URI, info hash, .torrent link — whatever the source handed over.
    pub torrent_id: Option<&'a str>,
    /// Info hash when known, used when the link itself is not resolvable.
    pub hash: Option<&'a str>,
    /// A debrid service is selected in settings.
    pub service_selected: bool,
    /// The service has an API key configured.
    pub service_ready: bool,
    /// The client has no network connection.
    pub offline: bool,
    pub mode: Option<DebridMode>,
}

/// The routing decision. `only` reports whether debrid only mode governs it,
/// so callers need not re-derive it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "lowercase")]
pub enum RouteDecision {
    Torrent { only: bool },
    Block { reason: BlockReason, only: bool },
    Resolve { id: String, only: bool },
}

/// A torrent id debrid can resolve: a magnet with a full btih hash, or a bare hex hash.
fn usable(torrent_id: Option<&str>) -> Option<&str> {
    let id = torrent_id?;
    let is_magnet = id.len() >= 7 && id[..7].eq_ignore_ascii_case("magnet:");
    if (is_magnet && parse_hash(id).is_some()) || (id.len() == 40 && parse_hash(id).is_some()) {
        Some(id)
    } else {
        None
    }
}

/// Decides how a play request should be handled.
pub fn route_debrid(input: &RouteInput) -> RouteDecision {
    // with no service selected debrid is entirely out of the picture, only mode included
    if !input.service_selected {
        return RouteDecision::Torrent { only: false };
    }
    let only = input.mode == Some(DebridMode::Only);
    if !input.service_ready {
        return if only {
            RouteDecision::Block { reason: BlockReason::Key, only }
        } else {
            RouteDecision::Torrent { only }
        };
    }
    if input.offline {
        return if only {
            RouteDecision::Block { reason: BlockReason::Offline, only }
        } else {
            RouteDecision::Torrent { only }
        };
    }
    match usable(input.torrent_id).or_else(|| usable(input.hash)) {
        Some(id) => RouteDecision::Resolve { id: id.to_string(), only },
        None => {
            if only {
                RouteDecision::Block { reason: BlockReason::Source, only }
            } else {
                RouteDecision::Torrent { only }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const HASH: &str = "0123456789abcdef0123456789abcdef01234567";

    fn base<'a>() -> RouteInput<'a> {
        RouteInput {
            torrent_id: Some(HASH),
            hash: None,
            service_selected: true,
            service_ready: true,
            offline: false,
            mode: Some(DebridMode::Prefer),
        }
    }

    #[test]
    fn no_service_selected_always_torrents() {
        let mut input = base();
        input.service_selected = false;
        input.mode = Some(DebridMode::Only);
        assert_eq!(route_debrid(&input), RouteDecision::Torrent { only: false });
    }

    #[test]
    fn missing_key_blocks_only_mode_and_torrents_prefer_mode() {
        let mut input = base();
        input.service_ready = false;
        assert_eq!(route_debrid(&input), RouteDecision::Torrent { only: false });
        input.mode = Some(DebridMode::Only);
        assert_eq!(route_debrid(&input), RouteDecision::Block { reason: BlockReason::Key, only: true });
    }

    #[test]
    fn offline_blocks_only_mode_and_torrents_prefer_mode() {
        let mut input = base();
        input.offline = true;
        assert_eq!(route_debrid(&input), RouteDecision::Torrent { only: false });
        input.mode = Some(DebridMode::Only);
        assert_eq!(route_debrid(&input), RouteDecision::Block { reason: BlockReason::Offline, only: true });
    }

    #[test]
    fn unusable_source_blocks_only_mode_and_torrents_prefer_mode() {
        let mut input = base();
        input.torrent_id = Some("https://example.com/file.torrent");
        assert_eq!(route_debrid(&input), RouteDecision::Torrent { only: false });
        input.mode = Some(DebridMode::Only);
        assert_eq!(route_debrid(&input), RouteDecision::Block { reason: BlockReason::Source, only: true });
    }

    #[test]
    fn resolvable_ids_resolve() {
        let input = base();
        assert_eq!(route_debrid(&input), RouteDecision::Resolve { id: HASH.into(), only: false });

        let magnet = format!("magnet:?xt=urn:btih:{HASH}");
        let mut input = base();
        input.torrent_id = Some(&magnet);
        input.mode = Some(DebridMode::Only);
        assert_eq!(route_debrid(&input), RouteDecision::Resolve { id: magnet.clone(), only: true });
    }

    #[test]
    fn falls_back_to_the_separate_hash_when_the_link_is_not_resolvable() {
        let mut input = base();
        input.torrent_id = Some("https://example.com/file.torrent");
        input.hash = Some(HASH);
        assert_eq!(route_debrid(&input), RouteDecision::Resolve { id: HASH.into(), only: false });
    }
}
